src/ripple/rpc/ShardArchiveHandler.h

```diff
@@ -117,23 +109,57 @@ class ShardArchiveHandler
 
     // Remove the archive being processed.
     void
-    remove(std::lock_guard<std::mutex>&);
+    remove(std::lock_guard<std::mutex> const&);
 
     void
     doRelease(std::lock_guard<std::mutex> const&);
 
-    static std::mutex instance_mutex_;
-    static pointer instance_;
+    bool
+    onClosureFailed(
+        std::string const& errorMsg,
+        std::lock_guard<std::mutex> const& lock);
+
+    bool
+    removeAndProceed(std::lock_guard<std::mutex> const& lock);
 
+    /////////////////////////////////////////////////
+    // m_ is used to protect access to downloader_,
+    // archives_, process_ and to protect setting and
+    // destroying sqliteDB_.
+    /////////////////////////////////////////////////
     std::mutex mutable m_;
+    std::unique_ptr<DatabaseDownloader> downloader_;
+    std::map<std::uint32_t, parsedURL> archives_;
+    bool process_;
+    std::unique_ptr<DatabaseCon> sqliteDB_;
+    /////////////////////////////////////////////////
+
     Application& app_;
     beast::Journal const j_;
-    std::unique_ptr<DatabaseCon> sqliteDB_;
-    std::shared_ptr<DatabaseDownloader> downloader_;
     boost::filesystem::path const downloadDir_;
     boost::asio::basic_waitable_timer<std::chrono::steady_clock> timer_;
-    bool process_;
-    std::map<std::uint32_t, parsedURL> archives_;
+    JobCounter jobCounter_;
+    TimerOpCounter timerCounter_;
+    ShardVerificationScheduler verificationScheduler_;
+};
+
+////////////////////////////////////////////////////////////////////
+// The RecoveryHandler is an empty class that is constructed by
+// the application when the ShardArchiveHandler's state database
+// is present at application start, indicating that the handler
+// needs to perform recovery. However, if recovery isn't needed
+// at application start, and the user subsequently submits a request
+// to download shards, we construct a ShardArchiveHandler rather
+// than a RecoveryHandler to process the request. With this approach,
+// type verification can be employed to determine whether the
+// ShardArchiveHandler was constructed in recovery mode by the
+// application, or as a response to a user submitting a request to
+// download shards.
+////////////////////////////////////////////////////////////////////
+class RecoveryHandler : public ShardArchiveHandler
+{
+public:
+    RecoveryHandler(Application& app, Stoppable& parent);
```

Comment:
How about a comment in the body of `RecoveryHandler` that says we'll need a virtual destructor if they make the class non-empty? I don't want to implement a thing that isn't currently needed.

Reply:
In terms of it being un-needed, I think what you're doing will work given our compilers.  But this FAQ (from isocpp.org) suggests it is also undefined behavior: https://isocpp.org/wiki/faq/virtual-functions#virtual-dtors  Note that the first three bullets in that FAQ exactly match the situation you're in.  Personally, I don't believe the minor reduction in efficiency is worth the risks associated with undefined behavior.

But read the FAQ for yourself and make your own choice.  If you choose not to have a virtual destructor, then a comment in the body of `RecoveryHandler` would be an improvement.  Putting in the virtual destructor would obviate the need for the comment.

Reply:
I looked it up, and I found one other [source](https://www.geeksforgeeks.org/virtual-destructor/) that confirms this undefined behavior. I will add the virtual destructor, but I find the undefined behavior strange because I've read **many** sources that say the destructor should be made virtual only when the base implements a virtual function. Anyway...

Reply:
All of the references that say the base needs to implement a virtual function in order to require a virtual destructor are, in fact, just plain wrong.

The place the virtual destructor becomes critically important is if the derived class adds any data members beyond what's in the base class.  It doesn't matter whether the base class has any virtuals or not.  Without the virtual destructor a call to the base class destructor will leave any data members added by the derived class undestroyed, potentially violating any number of postconditions.

Yes, I know your current implementation has no data members in the derived class.  That's why I was not _insisting_ that the problem be fixed.

Still, with current compilers, any form of undefined behavior is risky.  Thanks for fixing this bit of undefined behavior.